### PR TITLE
Fix "I'm feeling lucky" search to include all crates

### PR DESCRIPTION
**Pull request by syphar**
This fixes the _I'm feeling lucky_ search ( #789 ) to do an efficient search over all creates, not only the first 280 ones. 

I was searching and testing through many different ways to get a random record, and found this to be the best for our dataset. (see also [this stackoverflow answer](https://stackoverflow.com/a/8675160/1194456) for some overview). 

I tested this locally with a dummy-dataset with 600 crates >100 stars, and out of 10k of these queries there were only <50 which needed a second try. In the statement execution time was <10ms all the time. 

Other approaches felt too complex in code, were too slow, or were too slow randomly. 

I'm happy to implement any needed changes. 

In #789 @Kixiron was mentioning metrics to be added, while I'm not sure which metrics are needed.

### CODE REVIEW

File: src/web/releases.rs [hunk 1]
```diff
@@ -497,6 +497,69 @@ impl Default for Search {
...
+    // since there is a chance of this query returning an empty result,
+    // we retry a certain amount of times, and log a warning.
+    for _ in 0..20 {
+        let rows = ctry!(
+            req,
+            conn.query(
+                "WITH params AS (
+                    -- get maximum possible id-value in crates-table
+                    SELECT last_value AS max_id FROM crates_id_seq
+                )
```
Comment (jyn514) [inline src/web/releases.rs:510]: Don't CTEs inhibit optimizations? What do you think about inlining it to the one place it's used?

File: src/web/releases.rs [hunk 2]
```diff
@@ -497,6 +497,69 @@ impl Default for Search {
...
+                    crates.name,
+                    releases.version,
+                    releases.target_name
+                FROM (
+                    -- generate 500 random numbers in the ID-range. 
+                    -- this might have to be increased when we have to repeat 
+                    -- this query too often. 
+                    -- it depends on the percentage of crates with > 100 stars
+                    SELECT DISTINCT 1 + trunc(random() * params.max_id)::INTEGER AS id
+                    FROM params, generate_series(1, 500)
```
Comment (jyn514) [inline src/web/releases.rs:521]: Hard-coding 500 makes me a little nervous. Could we generate this live based on production info? Or does that slow down the query too much?

File: src/web/releases.rs [hunk 3]
```diff
@@ -497,6 +497,69 @@ impl Default for Search {
...
+                    target_name
+                )),
+            );
+
+            let mut resp = Response::with((status::Found, Redirect(url)));
+            resp.headers.set(Expires(HttpDate(time::now())));
+
+            return Ok(resp);
+        } else {
+            ::log::warn!("retrying random crate search");
```
Comment (jyn514) [inline src/web/releases.rs:555]:
```suggestion
            log::warn!("retrying random crate search");
```

File: src/web/releases.rs [hunk 4]
```diff
@@ -497,6 +497,69 @@ impl Default for Search {
...
+        }
+    }
+    ::log::error!("found no result in random crate search");
```
Comment (jyn514) [inline src/web/releases.rs:558]:
```suggestion
    log::error!("found no result in random crate search");
```

**Comment by syphar** [comment]
@jyn514 thanks for the review! I'll check the comments in detail in the evening today

**Comment by Kixiron** [comment]
The metrics were mostly because I was curious, I wanted to put a counter on the I'm feeling lucky button that reported how many times it was actually used

File: src/web/releases.rs [hunk 5]
```diff
@@ -497,6 +497,69 @@ impl Default for Search {
     }
 }
 
+fn redirect_to_random_crate(req: &Request, conn: &mut PoolClient) -> IronResult<Response> {
+    // since there is a chance of this query returning an empty result,
+    // we retry a certain amount of times, and log a warning.
+    for _ in 0..20 {
```
Comment (Kixiron) [inline src/web/releases.rs:503]: It seems strange to use a fallible query, is there not a way to make it always work?

File: src/web/releases.rs [hunk 6]
```diff
@@ -497,6 +497,69 @@ impl Default for Search {
...
+                    name,
+                    version,
+                    target_name
+                )),
+            );
+
+            let mut resp = Response::with((status::Found, Redirect(url)));
+            resp.headers.set(Expires(HttpDate(time::now())));
+
+            return Ok(resp);
```
Comment (Kixiron) [inline src/web/releases.rs:553]:
```suggestion
            log::debug!("finished random crate search on iteration {}", i);
            return Ok(resp);
```
Reporting how many tries each query took seems useful, you'll have to add `i` as the loop counter too

File: src/web/releases.rs (hunk 1 above)
Comment (syphar) [reply src/web/releases.rs:510]: CTEs can be also used as optimization fence when the optimizer doesn't build the best query plan, often developers know the data better. 

In this specific case it doesn't make a difference, so it's purely for readability (IMHO)

File: src/web/releases.rs (hunk 2 above)
Comment (syphar) [reply src/web/releases.rs:521]: generating that value live slows down the query too much, I valued speed higher.

Comment (syphar) [reply src/web/releases.rs:521]: the number depends on the ratio of crates >100 stars

File: src/web/releases.rs (hunk 5 above)
Comment (syphar) [reply src/web/releases.rs:503]: @Kixiron I tried (I think) 5-10 different approaches, the ones that always work are either too slow or involve things like a recursive CTE (which is in in the end similar to retrying here, while this is more readable IMHO).

**Comment by syphar** [comment]
> The metrics were mostly because I was curious, I wanted to put a counter on the I'm feeling lucky button that reported how many times it was actually used

@Kixiron wouldn't we get that from the access logs too? or don't we have these? 

I'll check other examples and see if it's easy to add

File: src/web/releases.rs (hunk 2 above)
Comment (syphar) [reply src/web/releases.rs:521]: while thinking about this, I have another idea (but not likely it's faster)

File: src/web/releases.rs (hunk 5 above)
Comment (syphar) [reply src/web/releases.rs]: I just ran some numbers: current query runs between 2 and 8 ms (~) 
and delivers a random crate directly in most cases (50 misses for 10k queries), 

the repetitions are more or less a safeguard. 

Also I could not test on a production dataset so I don't know the amount of gaps in the id-span on the crate-table (and the production github stars without backfilling everything locally)

**Comment by syphar** [comment]
@Kixiron @jyn514 I (hopefully) answered your questions and added two commits with the prosed log-statement changes and metrics. 
I admit I was definitely focusing on query speed, while sacrificing beauty or (probably a little) readability. 

When wanted I could likely produce something better readable, but slower ( ~100 or even 200 ms ) , 

or we could also just add the metrics to see if the feature is actually used or could be removed :)

File: src/web/releases.rs (hunk 2 above)
Comment (jyn514) [reply src/web/releases.rs:521]: Let me know if it works out :)

**Review by jyn514 — COMMENTED** [review]
> wouldn't we get that from the access logs too? or don't we have these?

The access logs are not integrated with the grafana logging, and also don't work when running locally. Plus I prefer to have as little depending on nginx as possible because it's not tested at all and doesn't go through review.

> or we could also just add the metrics to see if the feature is actually used or could be removed :)

I would be interested in that.

> When wanted I could likely produce something better readable, but slower ( ~100 or even 200 ms ) ,

I would prefer to have the faster query I think, the searches take long enough already. Thanks for offering though :)

> Also I could not test on a production dataset so I don't know the amount of gaps in the id-span on the crate-table (and the production github stars without backfilling everything locally)

Let me know how to test this, happy to runs some queries. I did run the 'redirect_to_random_query' once with `EXPLAIN ANALYZE` and it took less than a millisecond, great job :)

**Comment by syphar** [comment]
> > or we could also just add the metrics to see if the feature is actually used or could be removed :)
> 
> I would be interested in that.

To validate what I understand: I would create a separate PR only adding the metrics. Then we watch how often this is actually used and we decide to drop or fix? @jyn514 


> > Also I could not test on a production dataset so I don't know the amount of gaps in the id-span on the crate-table (and the production github stars without backfilling everything locally)
> 
> Let me know how to test this, happy to runs some queries. I did run the 'redirect_to_random_query' once with `EXPLAIN ANALYZE` and it took less than a millisecond, great job :)

I just wrote a small (ugly) tool: https://gist.github.com/syphar/949cdb2ee297ca5695bad179cfd99a6e  

With that you should be able to play around with the numbers :)

File: src/web/releases.rs (hunk 2 above)
Comment (syphar) [reply src/web/releases.rs:521]: didn't help 😢 was too slow :)

**Comment by jyn514** [comment]
> To validate what I understand: I would create a separate PR only adding the metrics. Then we watch how often this is actually used and we decide to drop or fix? @jyn514

I think this is a good change as-is, I would be happy merging now and we can add the metrics PR later.

> With that you should be able to play around with the numbers :)

Thanks! I tested this in prod and it never came up missing:
```
records: 10000
with result: 10000
without result: 0
```
So I don't think we have to worry.

Is this ready to merge? It looks good from my end :)

**Comment by syphar** [comment]
> > To validate what I understand: I would create a separate PR only adding the metrics. Then we watch how often this is actually used and we decide to drop or fix? @jyn514
> 
> I think this is a good change as-is, I would be happy merging now and we can add the metrics PR later.

After the comment by @Kixiron I already added metrics, as I understood how it would have to be done (see https://github.com/rust-lang/docs.rs/pull/1251/commits/e9bfc4721e65a0f90b1dc7b91a506ff8ad1a399f ) . 


> Is this ready to merge? It looks good from my end :)

From my side It would be ready, except we want more tests, different metrics, or any other change

**Comment by syphar** [comment]
@jyn514 just discovered by accident that my testing-gist used an even smaller number or random IDs (100) and still always delivered results. 

But 500 is fast enough, so I guess we're fine with the current number for some time.

**Pull request merged**